test/micro/org/openjdk/bench/java/lang/stable/CustomCachingPredicateBenchmark.java

```diff
@@ -0,0 +1,124 @@
+import org.openjdk.jmh.annotations.Threads;
+import org.openjdk.jmh.annotations.Warmup;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.LockSupport;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static org.openjdk.bench.java.lang.stable.CustomCachingFunctions.cachingPredicate;
+
+/**
+ * Benchmark measuring custom stable value types
+ */
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+@State(Scope.Benchmark) // Share the same state instance (for contention)
+@Warmup(iterations = 5, time = 1)
+@Measurement(iterations = 5, time = 2)
+@Fork(value = 2, jvmArgsAppend = {
+        "--enable-preview"
+        // Prevent the use of uncommon traps
+//        ,"-XX:PerMethodTrapLimit=0"
+})
+@Threads(Threads.MAX)   // Benchmark under contention
+//@OperationsPerInvocation(2)
+public class CustomCachingPredicateBenchmark {
+
+    private static final Set<Integer> SET = IntStream.range(0, 64).boxed().collect(Collectors.toSet());
+    private static final Predicate<Integer> EVEN = i -> {
+        // Slow down the original
+        LockSupport.parkNanos(TimeUnit.MILLISECONDS.toNanos(1));
+        return i % 2 == 0;
+    };
+
+    private static final Integer VALUE = 42;
+
+    private static final Predicate<Integer> PREDICATE = cachingPredicate(SET, EVEN);
+    private final Predicate<Integer> predicate = cachingPredicate(SET, EVEN);
+
+    @Benchmark
+    public boolean predicate() {
+        return predicate.test(VALUE);
+    }
+
+    @Benchmark
+    public boolean staticPredicate() {
+        return PREDICATE.test(VALUE);
+    }
+
+
+    //Benchmark                                        Mode  Cnt  Score   Error  Units
+    //CustomCachingPredicateBenchmark.predicate        avgt   10  3.054 ? 0.155  ns/op
+    //CustomCachingPredicateBenchmark.staticPredicate  avgt   10  2.205 ? 0.361  ns/op
+
+    // This is not constant foldable
+    record CachingPredicate<T>(Map<? extends T, StableValue<Boolean>> delegate,
+                               Predicate<T> original) implements Predicate<T> {
+
+        public CachingPredicate(Set<? extends T> inputs, Predicate<T> original) {
+            this(inputs.stream()
+                            .collect(Collectors.toMap(Function.identity(), _ -> StableValue.newInstance())),
+                    original
+            );
+        }
+
+        @Override
+        public boolean test(T t) {
+            final StableValue<Boolean> stable = delegate.get(t);
+            if (stable == null) {
+                throw new IllegalArgumentException(t.toString());
+
+            }
+            if (stable.isSet()) {
+                return stable.isSet();
+            }
+            synchronized (this) {
+                if (stable.isSet()) {
+                    return stable.isSet();
+                }
```

Comment:
`StableValue​::isSet()` would always return `true` after the first call to `test(…)`.
```suggestion
            if (stable.isSet()) {
                return stable.orElseThrow();
            }
            synchronized (this) {
                if (stable.isSet()) {
                    return stable.orElseThrow();
                }
```